provisioner/shell/provisioner.go

```diff
@@ -218,6 +228,57 @@ func (p *Provisioner) Provision(ui packer.Ui, comm packer.Communicator) error {
 		tf.Close()
 	}
 
+	if p.config.UseEnvVarFile == true {
+		tf, err := ioutil.TempFile("", "packer-shell-vars")
+		if err != nil {
+			return fmt.Errorf("Error preparing shell script: %s", err)
+		}
+		defer os.Remove(tf.Name())
+
+		// Write our contents to it
+		writer := bufio.NewWriter(tf)
+		if _, err := writer.WriteString(p.createEnvVarFileContent()); err != nil {
+			return fmt.Errorf("Error preparing shell script: %s", err)
+		}
+
+		if err := writer.Flush(); err != nil {
+			return fmt.Errorf("Error preparing shell script: %s", err)
+		}
+
+		p.config.envVarFile = tf.Name()
+		defer os.Remove(p.config.envVarFile)
+
+		// upload the var file
+		var cmd *packer.RemoteCmd
+		err = p.retryable(func() error {
+			if _, err := tf.Seek(0, 0); err != nil {
+				return err
+			}
+
+			var r io.Reader = tf
+			if !p.config.Binary {
+				r = &UnixReader{Reader: r}
+			}
+			remoteVFName := fmt.Sprintf("%s/%s", p.config.RemoteFolder, "varfile")
```

Comment:
🌵  I might be over cautious but may be using `tf.Name()` instead of `varfile` might reduce the risk of a collision.

I'm saying this because the script file had some random number in its name in the debug line after a try:
```
starting remote command: chmod +x /tmp/script_5423.sh; . /tmp/varfile && /tmp/script_5423.sh
```

Reply:
🗑 Also, a bit later I saw the debug line :
```
starting remote command: rm -f /tmp/script_5423.sh
```
May be we want to remove varfile too ?

Reply:
Sure, to both. We should only be uploading one varfile per provisioner but being extra careful doesn't hurt. And you're right that we should be removing the varfile from the guest system; I remembered to remove it from the host but forgot the guest.